### ISSUE TITLE
Probably, suggestions?

### ISSUE THREAD

**Issue by Gamer-XP** [mh-cz/GameMaker-Foreach#1]
Hey. Neat system. I was messing with it to make it a bit simpler. Ended up with following version (still pretty rough):

[foreach.txt](https://github.com/mh-cz/GameMaker-Foreach/files/9192586/foreach.txt)

Example of modified syntax:

`foreach i in 1, 5, 0.5 run`
`foreach c in "STRING" run`
`foreach elem in some_array run`
`foreach elem in some_struct run`
`foreach item in some_ds_list as_list run`

Basically, I merged all "in*" macro into single one that automatically selects what enumerator should be used if possible. If not - there are "as_*" additional macro you can add before run to define the type of variable.

Just sharing it here, as reference for different implementation.

**Comment by mh-cz** [comment]
(sry for such late response)
I think this is a nice addition. Not having to specify the basic types was something I originally went for and I had it before and then the DS types fcked it up. There's just one thing about the syntax. I can't really use the lowercase "run" because it's an extremely common variable name. Otherwise I'd leave everything lowercase too. Maybe we can find another short word to replace the "run"? "exec" doesn't sound that bad. Or literally just "_"
`foreach elem in something exec { ... }`
`foreach elem in something _ { ... }`

**Comment by Gamer-XP** [comment]
I think I had lowercase issue with few other variables as well after I renamed them.. I think it was Loop. 
So yeah, lowercase version is a bit hard to make universal - I just made it for myself since I'm fine with editing my project a bit to make it work. 
"run" and "loops" are really likely to be found in already existing projects, yeah. I'd not use "_" since it looks a bit weird. "exec" may be fine I guess? "loop" should be changed to something else too, like... "forvar", so you can do stuff like "forvar.get()", "forvar.set( 123 )"

Also, there is "final" version I'm using now. I did more renamings and added function to register custom struct-type as something you can use with foreach. I'm not sure - maybe it's using some of my utility functions from other scripts though.

[foreach.txt](https://github.com/mh-cz/GameMaker-Foreach/files/9571646/foreach.txt)

**Comment by mh-cz** [comment]
ok so I went through your code, simplified it and optimized it a bit since it was running around 500 ms slower in my benchmarks
"map" is now "set", "run" is now "exec" and "Loop" is now "fe" (I used "fe" in older versions and it didn't cause any trouble)
here's the current syntax and I like it so far
```
foreach v in 4 exec list[| v] = (v+1)*100;
foreach v in grid as_grid exec fe.set([fe.xpos+1, fe.ypos+1]);
foreach v in arr exec fe.set(fe.i);

foreach v in list as_list exec ...
foreach v in grid as_grid exec ...
foreach v in struct exec ...
foreach v in arr exec ...

foreach v in 5 exec ...
foreach v in -5, 5 exec ...
foreach v in -5, 5, 2 exec ...
```
here's the code
[foreach.txt](https://github.com/mh-cz/GameMaker-Foreach/files/9574554/foreach.txt)

**Comment by Gamer-XP** [comment]
It's a bit hard to compare when they're that different, but for optimization - did you remove nested Enumerator and made Auto-Detect into a function? Well, I guess there were no much reason for it to be an enumerator. I think I made it as once because I made AutoDetect do something itself in the beginning, but stopped doing it later.

**Comment by mh-cz** [comment]
God I wish yoyo made ds types somehow distinguishable. I was trying to find a workaround before. The only way would be to make your own ds handler that tracks the types for you. But no one would want to use that tho. Wonder what went through their heads to make it this way

**Comment by Gamer-XP** [comment]
I think ds stuff was there even before YoYo, so it's Mark Overmars who wrote that part. It was all numbers, no references at all. Probably, was written that way to make it easier to understand for newbie programmers?

I think they said they're going to change it in the future, like they changed already how object references work.

**Comment by mh-cz** [comment]
Aight I'm gonna close this one. You can msg me discord anytime
